### DISCUSSION TITLE
ForEach array not collecting correctly

### DISCUSSION THREAD

**Discussion by Kraziken** [PowerShell/PowerShell#16082] [Q&A]
I have this small command that worked fine until I updated to the latest powershell 7 build.  

Any idea why the array collection would not work?  After it completes, the $onlinecomputers and $offlinecomputers are a single variable with all the computer names mashed together instead of an array.

I ran this on a machine that had not been updated to Powershell 7, and the array collections worked fine.

[system.collections.arraylist]$Onlinecomputers=@()
[system.collections.arraylist]$Offlinecomputers=@()

ForEach ($server in $servers)
    {
    if (Test-Connection -ComputerName $server -Count 1 -Quiet)
        {
        Write-Host $server is online
        $onlinecomputers += $server
        }
    Else
        {
        Write-Host $server is offline
        $offlinecomputers += $server
        }
    }

**Comment by vexx32** [comment]
Not really sure what you're seeing here. It works as expected for me:

```ps1
$servers = 'google.com', 'facebook.com', 'github.com'
foreach ($server in $servers) {
    if (Test-Connection $server -Quiet -Count 1) {
        write-host "$server is online"
        $online += $server
    }
    else {
        write-host "$server is offline"
        $offline += $server
    }
}
```
```code
google.com is online
facebook.com is online
github.com is online
```
```ps1
PS> $online
```
```code
google.com
facebook.com
github.com
```

I will note that since you've used `+=` to add things to your collections, your `ArrayList` cast is... doing a lot of extra array copies. When you `+=`, PowerShell quietly turns your `ArrayList` into a basic `Array`, and then because of the variable type cast, it converts it back into a _new_ `ArrayList` again. So you're generating `$servers.Count` _new_ arrays and new arraylists every single time you run the foreach loop, which is kind of wasteful.

If you use `$offlineComputers.Add($server)` (and similar for online) instead, you can keep the one arraylist instead of making several. Since you're using arraylist, though, you'll need to silence the output from the `.Add()` method with something like `$offlineComputers.Add($server) > $null`. Alternatively you can use `[System.Collections.Generic.List[object]]` and use `.Add()` without worrying about extra output, because it doesn't have any.
